File: mobile-apps/ios-node-app/MobileGradientPayloadAdapter.swift
import Foundation

struct CanonicalGradientPayload {
    let round: Int
    let nodeID: String
    let modelHashHex: String
    let gradientChunkB64: String
    let timestampUnixMS: Int64

    func canonicalBytes() -> Data {
        let canonical = "round=\(round)\nnode_id=\(nodeID)\nmodel_hash=\(modelHashHex)\ngradient_chunk_b64=\(gradientChunkB64)\ntimestamp_unix_ms=\(timestampUnixMS)"
        return Data(canonical.utf8)
    }
}

struct SignedGradientEnvelope {
    let nodeID: String
    let round: Int
    let signerAlias: String
    let publicKeyPEM: String
    let gradientPayloadB64: String
    let gradientSignatureB64: String
    let attestationPayloadB64: String

    func asJSONObject() -> [String: Any] {
        [
            "node_id": nodeID,
            "round": round,
            "signer_alias": signerAlias,
            "public_key_pem": publicKeyPEM,
            "gradient_payload_b64": gradientPayloadB64,
            "gradient_signature_b64": gradientSignatureB64,
            "attestation_payload_b64": attestationPayloadB64,
        ]
    }

    func asJSONString() -> String {
        guard let data = try? JSONSerialization.data(withJSONObject: asJSONObject(), options: [.sortedKeys]) else {
            return "{}"
        }
        return String(data: data, encoding: .utf8) ?? "{}"
    }
}

enum MobilePayloadAdapter {
    static func buildSignedEnvelope(
        nodeID: String,
        round: Int,
        modelHashHex: String,
        gradientChunk: Data,
        signerAlias: String,
        signer: MobileHardwareSigner
    ) throws -> SignedGradientEnvelope {
        let payload = CanonicalGradientPayload(
            round: round,
            nodeID: nodeID,
            modelHashHex: modelHashHex,
            gradientChunkB64: gradientChunk.base64EncodedString(),
            timestampUnixMS: Int64(Date().timeIntervalSince1970 * 1000)
        )
        let canonicalBytes = payload.canonicalBytes()
        let signature = try signer.sign(alias: signerAlias, payload: canonicalBytes)
        let attestation = try signer.attestation(alias: signerAlias)
        let publicKeyPEM = try signer.publicKeyPEM(alias: signerAlias)

        return SignedGradientEnvelope(
            nodeID: nodeID,
            round: round,
            signerAlias: signerAlias,
            publicKeyPEM: publicKeyPEM,
            gradientPayloadB64: canonicalBytes.base64EncodedString(),
            gradientSignatureB64: signature.base64EncodedString(),
            attestationPayloadB64: attestation.base64EncodedString()
        )
    }
}

File: mobile-apps/android-node-app/MobileGradientPayloadAdapter.kt
package io.sovereignmap.node

import android.util.Base64
import org.json.JSONObject
import java.nio.charset.StandardCharsets

object MobileGradientPayloadAdapter {
    data class CanonicalGradientPayload(
        val round: Int,
        val nodeId: String,
        val modelHashHex: String,
        val gradientChunkB64: String,
        val timestampUnixMs: Long,
    ) {
        fun canonicalBytes(): ByteArray {
            val canonical = "round=$round\n" +
                "node_id=$nodeId\n" +
                "model_hash=$modelHashHex\n" +
                "gradient_chunk_b64=$gradientChunkB64\n" +
                "timestamp_unix_ms=$timestampUnixMs"
            return canonical.toByteArray(StandardCharsets.UTF_8)
        }
    }

    data class SignedGradientEnvelope(
        val nodeId: String,
        val round: Int,
        val signerAlias: String,
        val publicKeyPem: String,
        val gradientPayloadB64: String,
        val gradientSignatureB64: String,
        val attestationPayloadB64: String,
    ) {
        fun asJson(): JSONObject {
            return JSONObject()
                .put("node_id", nodeId)
                .put("round", round)
                .put("signer_alias", signerAlias)
                .put("public_key_pem", publicKeyPem)
                .put("gradient_payload_b64", gradientPayloadB64)
                .put("gradient_signature_b64", gradientSignatureB64)
                .put("attestation_payload_b64", attestationPayloadB64)
        }
    }

    fun buildSignedEnvelope(
        nodeId: String,
        round: Int,
        modelHashHex: String,
        gradientChunk: ByteArray,
        signerAlias: String,
        signer: MobileHardwareSigner,
    ): SignedGradientEnvelope {
        val payload = CanonicalGradientPayload(
            round = round,
            nodeId = nodeId,
            modelHashHex = modelHashHex,
            gradientChunkB64 = Base64.encodeToString(gradientChunk, Base64.NO_WRAP),
            timestampUnixMs = System.currentTimeMillis(),
        )

        val canonicalBytes = payload.canonicalBytes()
        val signature = signer.sign(signerAlias, canonicalBytes)
        val attestation = signer.attestation(signerAlias)
        val publicKeyPem = signer.publicKeyPEM(signerAlias)

        return SignedGradientEnvelope(
            nodeId = nodeId,
            round = round,
            signerAlias = signerAlias,
            publicKeyPem = publicKeyPem,
            gradientPayloadB64 = Base64.encodeToString(canonicalBytes, Base64.NO_WRAP),
            gradientSignatureB64 = Base64.encodeToString(signature, Base64.NO_WRAP),
            attestationPayloadB64 = Base64.encodeToString(attestation, Base64.NO_WRAP),
        )
    }
}
